lib/cache/add-remote-git.js

```diff
@@ -173,97 +176,124 @@ function fetchRemote (p, u, co, origUrl, cb) {
   )
 }
 
-function resolveHead (p, u, co, origUrl, cb) {
+function resolveHead (params, cb) {
+  log.verbose('resolveHead', 'original treeish', params.treeish)
   git.whichAndExec(
-    [ "rev-list", "-n1", co ],
-    { cwd : p, env : gitEnv() },
+    ['rev-list', '-n1', params.treeish],
+    { cwd: params.checkoutPath, env: gitEnv() },
     function (er, stdout, stderr) {
-      stdout = (stdout + "\n" + stderr).trim()
+      params.exactTreeish = (stdout + '\n' + stderr).trim()
       if (er) {
-        log.error("Failed resolving git HEAD (" + u + ")", stderr)
+        log.error('Failed resolving git HEAD (' + params.gitURL + ')', stderr)
         return cb(er)
       }
-      log.verbose("git rev-list -n1 " + co, stdout)
-      var parsed = url.parse(origUrl)
-      parsed.hash = stdout
-      var resolved = url.format(parsed)
+      log.verbose('resolveHead', 'resolved treeish', params.exactTreeish)
 
-      if (!/^git[+:]/.test(parsed.protocol)) {
-        resolved = "git+" + resolved
-      }
+      params.resolved = getResolved(params.originalURL, params.exactTreeish)
 
-      // https://github.com/npm/npm/issues/3224
-      // node incorrectly sticks a / at the start of the path We know that the
-      // host won't change, so split and detect this
-      var spo = origUrl.split(parsed.host)
-      var spr = resolved.split(parsed.host)
-      if (spo[1].charAt(0) === ":" && spr[1].charAt(0) === "/") {
-        spr[1] = spr[1].slice(1)
-      }
-      resolved = spr.join(parsed.host)
+      // generate a unique filename
+      params.tmpdir = path.join(
+        npm.tmp,
+        'git-cache-' + randomBytes(6).toString('hex'),
+        params.exactTreeish
+      )
+      log.silly('resolveHead', 'temporary clone directory', params.tmpdir)
 
-      log.verbose("resolved git url", resolved)
-      cache(p, u, stdout, resolved, cb)
+      mkdir(params.tmpdir, function (er) {
+        if (er) return cb(er)
+
+        clone(params, cb)
+      })
     }
   )
 }
 
 /**
- * Make an actual clone from the bare (mirrored) cache. There is no safe way to
- * do a one-step clone to a treeish that isn't guaranteed to be a branch, so
- * this has to be two steps.
+ * Make an actual clone from the bare (mirrored) cache.
  */
-function cache (p, u, treeish, resolved, cb) {
-  // generate a unique filename
-  randomBytes(6, function (er, random) {
-    if (er) return cb(er)
-
-    var tmp = path.join(
-      npm.tmp,
-      "git-cache-"+random.toString("hex"),
-      treeish
-    )
-
-    mkdir(tmp, function (er) {
-      if (er) return cb(er)
-
-      git.whichAndExec(["clone", p, tmp], { cwd : p, env : gitEnv() }, clone)
-    })
-
-    function clone (er, stdout, stderr) {
-      stdout = (stdout + "\n" + stderr).trim()
+function clone (params, cb) {
```

Comment:
I'll brainstorm this as I rewrite. More descriptive function names would probably be a good complement to the variable names.